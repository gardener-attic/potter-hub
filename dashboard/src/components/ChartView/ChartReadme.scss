.ChartReadme--error {
  flex-direction: column;
  height: 100%;
}

.ChartReadme p {
  margin: 0.625em 0;
}

.ChartReadme pre {
  margin: 0.625em 0;
  border-radius: 5px;
  overflow: auto;
}

.ChartReadme blockquote {
  margin: 0;
  text-align: left;
  font-size: 1em;
  border-radius: 5px;
  border: 1px solid #ddd;
  background-color: #f1f1f1;
  padding: 0.1em 1em;
  max-width: 100%;
}

.ChartReadme blockquote:before,
.ChartReadme blockquote:after {
  background-image: none;
}

.ChartReadme tr {
  border-bottom: 1px solid #eee;
}
