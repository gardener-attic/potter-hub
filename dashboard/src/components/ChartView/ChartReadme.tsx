import * as React from "react";
import { FileText } from "react-feather";
import ReactMarkdown from "react-markdown";

import LoadingWrapper from "../LoadingWrapper";
import HeadingRenderer from "./HeadingRenderer";

import "./ChartReadme.css";
import LinkRenderer from "./LinkRenderer";

interface IChartReadmeProps {
  getChartReadme: (version: string) => void;
  hasError: boolean;
  readme?: string;
  version: string;
}

class ChartReadme extends React.Component<IChartReadmeProps> {
  public componentDidMount() {
    const { getChartReadme, version } = this.props;
    getChartReadme(version);
  }

  public componentDidUpdate(prevProps: IChartReadmeProps) {
    const { getChartReadme, version } = this.props;
    if (version !== prevProps.version) {
      getChartReadme(version);
    }
  }

  public render() {
    const { hasError, readme } = this.props;
    if (hasError) {
      return this.renderError();
    }
    return (
      <LoadingWrapper loaded={!!readme}>
        {readme && (
          <div className="ChartReadme">
            <ReactMarkdown
              source={readme}
              renderers={{
                heading: HeadingRenderer,
                link: LinkRenderer,
              }}
              skipHtml={true}
            />
          </div>
        )}
      </LoadingWrapper>
    );
  }

  public renderError() {
    return (
      <div className="ChartReadme ChartReadme--error flex align-center text-c">
        <FileText size={64} />
        <h4>No README found</h4>
      </div>
    );
  }
}

export default ChartReadme;
