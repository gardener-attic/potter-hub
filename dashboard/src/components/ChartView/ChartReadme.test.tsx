import { mount, shallow } from "enzyme";
import context from "jest-plugin-context";
import * as React from "react";
import ReactMarkdown from "react-markdown";
import { BrowserRouter } from "react-router-dom";
import { HashLink as Link } from "react-router-hash-link";

import itBehavesLike from "../../shared/specs";

import ChartReadme from "./ChartReadme";

context("when readme is not present", () => {
  itBehavesLike("aLoadingComponent", {
    component: ChartReadme,
    props: {
      getChartReadme: jest.fn(),
      hasError: false,
      version: "1.2.3",
    },
  });
});

describe("getChartReadme", () => {
  const spy = jest.fn();
  const wrapper = shallow(<ChartReadme getChartReadme={spy} hasError={false} version="1.2.3" />);

  it("gets triggered when mounting", () => {
    expect(spy).toHaveBeenCalledWith("1.2.3");
  });

  it("gets triggered after changing version", () => {
    wrapper.setProps({ version: "1.2.4" });
    expect(spy).toHaveBeenCalledWith("1.2.4");
  });

  it("does not get triggered when version doesn't change", () => {
    wrapper.setProps({ version: "1.2.4" });
    wrapper.setProps({ hasError: true });
    expect(spy).toHaveBeenCalledTimes(2);
  });
});

it("renders the ReactMarkdown content is readme is present", () => {
  const wrapper = shallow(
    <ChartReadme
      getChartReadme={jest.fn()}
      hasError={false}
      version="1.2.3"
      readme="# Markdown Readme"
    />,
  );
  const component = wrapper.find(ReactMarkdown);
  expect(component.props()).toMatchObject({ source: "# Markdown Readme" });
});

it("renders an error when hasError is set", () => {
  const wrapper = shallow(
    <ChartReadme getChartReadme={jest.fn()} hasError={true} version="1.2.3" />,
  );
  expect(wrapper.text()).toContain("No README found");
});

it("renders the ReactMarkdown content adding IDs for the titles", () => {
  const wrapper = mount(
    <ChartReadme
      getChartReadme={jest.fn()}
      hasError={false}
      version="1.2.3"
      readme="# _Markdown_ 'Readme_or_not'!"
    />,
  );
  const component = wrapper.find("#markdown-readme_or_not");
  expect(component).toExist();
});

it("renders the ReactMarkdown ignoring comments", () => {
  const wrapper = mount(
    <ChartReadme
      getChartReadme={jest.fn()}
      hasError={false}
      version="1.2.3"
      readme={`<!-- This is a comment -->
      This is text`}
    />,
  );
  const html = wrapper.html();
  expect(html).toContain("This is text");
  expect(html).not.toContain("This is a comment");
});

it("renders the ReactMarkdown content with hash links", () => {
  const wrapper = mount(
    <BrowserRouter>
      <ChartReadme
        getChartReadme={jest.fn()}
        hasError={false}
        version="1.2.3"
        readme={`[section 1](#section-1)
      # Section 1`}
      />
    </BrowserRouter>,
  );
  expect(wrapper.find(Link)).toExist();
});
