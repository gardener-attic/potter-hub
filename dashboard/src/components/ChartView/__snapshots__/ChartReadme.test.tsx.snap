// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when readme is not present loading spinner matches the snapshot 1`] = `
<LoadingWrapper
  loaded={false}
  type={0}
/>
`;
